storage2/src/storage.rs

```diff
@@ -0,0 +1,227 @@
+    /// `Storage`, or `None` if the tree is empty.
+    pub async fn latest_version(&self) -> Result<Option<jmt::Version>> {
+        Ok(self
+            .get_rightmost_leaf()
+            .await?
+            .map(|(node_key, _)| node_key.version()))
+    }
+
+    /// Returns a new [`State`] on top of the latest version of the tree.
+    pub async fn state(&self) -> Result<State> {
+        // If the tree is empty, use PRE_GENESIS_VERSION as the version,
+        // so that the first commit will be at version 0.
+        let version = self
+            .latest_version()
+            .await?
+            .unwrap_or(WriteOverlay::<Storage>::PRE_GENESIS_VERSION);
+
+        tracing::debug!("creating state for version {}", version);
+        Ok(Arc::new(RwLock::new(WriteOverlay::new(
+            self.clone(),
+            version,
+        ))))
+    }
+
+    /// Like [`Self::state`], but bundles in a [`tonic`] error conversion.
+    ///
+    /// This is useful for implementing gRPC services that query the storage:
+    /// each gRPC request can create an ephemeral [`State`] pinning the current
+    /// version at the time the request was received, and then query it using
+    /// component `View`s to handle the request.
+    pub async fn state_tonic(&self) -> std::result::Result<State, tonic::Status> {
+        self.state()
+            .await
+            .map_err(|e| tonic::Status::internal(e.to_string()))
+    }
+
+    pub async fn put_nct(&self, tct: &tct::Tree) -> Result<()> {
+        let db = self.0.clone();
+
+        tracing::debug!("serializing TCT");
+        let tct_data = bincode::serialize(tct)?;
+        tracing::debug!(tct_bytes = tct_data.len(), "serialized TCT");
+        gauge!(metrics::TCT_SIZE_BYTES, tct_data.len() as f64);
+
+        let span = Span::current();
+        tokio::task::Builder::new()
+            .name("put_nct")
+            .spawn_blocking(move || {
+                span.in_scope(|| {
+                    let nct_cf = db.cf_handle("nct").expect("nct column family not found");
+                    db.put_cf(nct_cf, "tct", &tct_data)?;
+                    Ok::<_, anyhow::Error>(())
+                })
+            })
+            .unwrap()
+            .await?
+    }
+
+    pub async fn get_nct(&self) -> Result<tct::Tree> {
+        let db = self.0.clone();
+        let span = Span::current();
+        tokio::task::Builder::new()
+            .name("get_nct")
+            .spawn_blocking(move || {
+                span.in_scope(|| {
+                    let nct_cf = db.cf_handle("nct").expect("nct column family not found");
+                    if let Some(tct_bytes) = db.get_cf(nct_cf, "tct")? {
+                        Ok(bincode::deserialize(&tct_bytes)?)
+                    } else {
+                        Ok(tct::Tree::new())
+                    }
+                })
+            })
+            .unwrap()
+            .await?
+    }
```

Comment:
These methods violate the design principles of the rest of the API (e.g., they aren't versioned, and they can screw up someone else's view of the `State`), so we should delete them, and figure out how to record the NCT through the new API we're designing.